### PR TITLE
Feature/us5183 leader sends individuals messages

### PR DESCRIPTION
**Pull request by Doug-Shannon**
leader sends individual message + only leader can send group

### CODE REVIEW

File: Gateway/MinistryPlatform.Translation/Models/MpGroupParticipant.cs [hunk 1]
```diff
@@ -17,10 +17,17 @@ public class MpGroupParticipant
...
         public int ContactId { get; set; }
+        [JsonProperty("Nickname")]
         public string NickName { get; set; }
+        [JsonProperty("Display_Name")]
+        public string DisplayName { get; set; }
+        [JsonProperty("Last_name")]
         public string LastName { get; set; }
+        [JsonProperty("Role_Title")]
         public string GroupRoleTitle { get; set; }
+        [JsonProperty("email_address")]
```
Comment (JoeKer1) [inline Gateway/MinistryPlatform.Translation/Models/MpGroupParticipant.cs]: We may want to make sure everything matches MP including case (comment stolen from @andycanterbury )

File: Gateway/MinistryPlatform.Translation/Repositories/GroupParticipantRepository.cs [hunk 2]
```diff
@@ -115,12 +116,13 @@ public int GetRsvpYesCount(int groupId, int eventId)
 
         public List<MpGroup> GetAllGroupNamesLeadByParticipant(int participantId, int groupType = -1)
```
Comment (JoeKer1) [inline Gateway/MinistryPlatform.Translation/Repositories/GroupParticipantRepository.cs]: should we make it int? groupType = null?

File: Gateway/MinistryPlatform.Translation/Repositories/GroupParticipantRepository.cs [hunk 3]
```diff
@@ -144,13 +146,57 @@ public bool GetIsLeader(int participantId, int groupType = -1)
...
+            {
+                search += $" AND Group_ID_Table.Group_Type_ID = {groupType}";
             }
 
             var mpGroupParticipants = _ministryPlatformRest.UsingAuthenticationToken(_apiUserService.GetToken()).Search<MpGroupParticipant>(search, COLUMNS);
 
             return mpGroupParticipants.Any();
         }
+
+        public List<MpGroupParticipant> GetAllParticipantsForLeaderGroups(int participantId, int? groupType = -1, int? groupId = -1)
```
Comment (JoeKer1) [inline Gateway/MinistryPlatform.Translation/Repositories/GroupParticipantRepository.cs]: int? groupType = null and int? groupId = null

Would have to update the ifs as well.

File: Gateway/MinistryPlatform.Translation/Repositories/Interfaces/IMinistryPlatformRestRepository.cs [hunk 4]
```diff
@@ -47,8 +47,10 @@ public interface IMinistryPlatformRestRepository
         /// <typeparam name="T">The type of record to get.  This should correspond to an appropriately annotated model class, so that MP columns can be properly mapped (using NewtonSoft.Json) from MP to the model object.  The model class must also be annotated with the RestApiTable attribute, specifying the actual MP table name.</typeparam>
         /// <param name="searchString">An "MP SQL" WHERE clause, for instance "Payment_Type_Id > 5 AND Payment_Type_Id &lt; 9".  If not specified, all rows will be returned.</param>
         /// <param name="selectColumns">Optionally specify which columns to retrieve from MP.  This is a comma-separated list of column names.  If not specified, all columns will be retrieved.</param>
+        /// <param name="orderByString">Optionally specify how to order retrieved columns. </param>
+        /// <param name="selectColumns">Specify whether only distinct records should be returned.  defaults to false.</param>
```
Comment (JoeKer1) [inline Gateway/MinistryPlatform.Translation/Repositories/Interfaces/IMinistryPlatformRestRepository.cs]: add distinct param comment (stolen from @andycanterbury's review!)

File: Gateway/MinistryPlatform.Translation/Repositories/Interfaces/IMinistryPlatformRestRepository.cs [hunk 5]
```diff
@@ -57,7 +59,7 @@ public interface IMinistryPlatformRestRepository
         /// <param name="searchString">An "MP SQL" WHERE clause, for instance "Payment_Type_Id > 5 AND Payment_Type_Id &lt; 9".</param>
         /// <param name="columns">Optionally specify which columns to retrieve from MP.  This is a comma-separated list of column names.</param>
         /// <returns>An List of objects representing the matching MP rows for the search, if found.</returns>
```
Comment (JoeKer1) [inline Gateway/MinistryPlatform.Translation/Repositories/Interfaces/IMinistryPlatformRestRepository.cs]: Do we need to add comments here as well? Prolly.

File: Gateway/MinistryPlatform.Translation/Repositories/MinistryPlatformRestRepository.cs [hunk 6]
```diff
@@ -194,11 +194,15 @@ private static string FormatStoredProcParameters(Dictionary<string, object> para
             return result.TrimEnd('&');
         }
 
-        public List<T> Search<T>(string searchString = null, string selectColumns = null)
+        public List<T> Search<T>(string searchString = null, string selectColumns = null, string orderByString = null, bool distinct = false)
         {
             var search = string.IsNullOrWhiteSpace(searchString) ? string.Empty : $"?$filter={MpRestEncode(searchString)}";
+            ////////////////////////////////////////////////////////////////////////This is why we became programmers right?
```
Comment (JoeKer1) [inline Gateway/MinistryPlatform.Translation/Repositories/MinistryPlatformRestRepository.cs]: lol probably remove this sword comment. 
      /| ________________
O|===|* >________________>
      \|

Comment (JoeKer1) [inline Gateway/MinistryPlatform.Translation/Repositories/MinistryPlatformRestRepository.cs]: ASCII art sucks in github :(

File: crossroads.net/app/my_serve/serve_team.service.js [hunk 7]
```diff
@@ -42,6 +29,10 @@ export default class ServeTeamService {
         }, message).$promise;
     }
 
+    sendParticipantsMessage(message) {
+        return this.resource(__API_ENDPOINT__ + 'api/group/messageselectparticipants').save(message).$promise;
```
Comment (JoeKer1) [inline crossroads.net/app/my_serve/serve_team.service.js]: Use string interpolation

File: crossroads.net/app/my_serve/serve_team_message/serveTeamMessage.controller.js [hunk 8]
```diff
@@ -48,17 +54,33 @@ export default class ServeTeamMessageController {
...
-        this.rootScope.$emit('notify', this.rootScope.MESSAGES.emailSent);
-        this.state.go('serve-signup');
-      })
-      .catch((err) => {
-        this.rootScope.$emit('notify', this.rootScope.MESSAGES.messageSendError);
-      })
-      .finally(() => {
-        this.processing = false;
-      });
+<<<<<<< HEAD
```
Comment (JoeKer1) [inline crossroads.net/app/my_serve/serve_team_message/serveTeamMessage.controller.js]: Merge conflict?

File: crossroads.net/app/my_serve/serve_team.service.js (hunk 7 above)
Comment (Doug-Shannon) [inline crossroads.net/app/my_serve/serve_team.service.js]: I don't think this one really matters, it is similar to how we are doing it elsewhere.

**Comment by Doug-Shannon** [comment]
Changes made and checked in.

File: crossroads.net/app/my_serve/serve_team_message/serveTeamMessage.controller.js (hunk 8 above)
Comment (Doug-Shannon) [inline crossroads.net/app/my_serve/serve_team_message/serveTeamMessage.controller.js]: fixed

File: Gateway/MinistryPlatform.Translation/Repositories/MinistryPlatformRestRepository.cs (hunk 6 above)
Comment (Doug-Shannon) [inline Gateway/MinistryPlatform.Translation/Repositories/MinistryPlatformRestRepository.cs]: removed, but also lame.

File: Gateway/MinistryPlatform.Translation/Repositories/Interfaces/IMinistryPlatformRestRepository.cs (hunk 4 above)
Comment (Doug-Shannon) [inline Gateway/MinistryPlatform.Translation/Repositories/Interfaces/IMinistryPlatformRestRepository.cs]: fixed

File: Gateway/MinistryPlatform.Translation/Repositories/GroupParticipantRepository.cs (hunk 3 above)
Comment (Doug-Shannon) [inline Gateway/MinistryPlatform.Translation/Repositories/GroupParticipantRepository.cs]: fixed

File: Gateway/MinistryPlatform.Translation/Repositories/GroupParticipantRepository.cs (hunk 2 above)
Comment (Doug-Shannon) [inline Gateway/MinistryPlatform.Translation/Repositories/GroupParticipantRepository.cs]: fixed

File: Gateway/MinistryPlatform.Translation/Models/MpGroupParticipant.cs (hunk 1 above)
Comment (Doug-Shannon) [inline Gateway/MinistryPlatform.Translation/Models/MpGroupParticipant.cs]: fixed

**Pull request merged**